scanpointgenerator/core/dimension.py

```diff
@@ -81,119 +101,107 @@ def get_mesh_map(self, axis):
+        if hasattr(gen, "bounds") and gen.bounds:
+            tilings = reduce(lambda v, gen: v * gen.size, self.generators[:-1], 1)
+            if gen.alternate:
+                tilings /= 2.
+            for axis in gen.axes:
+                upper_base = gen.bounds[axis][1:]
+                lower_base = gen.bounds[axis][:-1]
+                upper, lower = upper_base, lower_base
+                if gen.alternate:
+                    upper = np.append(upper_base, lower_base[::-1])
+                    lower = np.append(lower_base, upper_base[::-1])
+                upper = np.tile(upper, int(tilings))
+                lower = np.tile(lower, int(tilings))
+                if tilings % 1 != 0:
+                    upper = np.append(upper, upper_base)
+                    lower = np.append(lower, lower_base)
+                axis_bounds_upper[axis] = upper
+                axis_bounds_lower[axis] = lower
+
         self.mask = mask
         self.indices = self.mask.nonzero()[0]
         self.size = len(self.indices)
+        self.positions = {axis:axis_positions[axis][self.indices] for axis in axis_positions}
+        self.upper_bounds = {axis:self.positions[axis] for axis in self.positions}
+        self.lower_bounds = {axis:self.positions[axis] for axis in self.positions}
+        for axis in axis_bounds_lower:
+            self.upper_bounds[axis] = axis_bounds_upper[axis][self.indices]
+            self.lower_bounds[axis] = axis_bounds_lower[axis][self.indices]
         self._prepared = True
 
+
     @staticmethod
     def merge_dimensions(dimensions):
-        """Merge multiple dimensions into one, scaling structures as required
-
-        Args:
-            dimensions (list): dimensions to merge (outermost first)
-        Returns:
-            Dimension: squashed dimension
-        """
-        final_dim = Dimension(dimensions[0].generators[0])
-        final_dim.generators = []
-        final_dim.lower = []
-        final_dim.upper = []
-        final_dim.axes = []
-        final_dim._max_length = 1
-        # masks in the inner generator are tiled by the size of
-        # outer generators and outer generators have their elements
-        # repeated by the size of inner generators
-        for dim in dimensions:
-            inner_masks = [m.copy() for m in dim._masks] # copy masks to preserve input strucutres
-            for m in final_dim._masks:
-                m["repeat"] *= dim._max_length
-            for m in inner_masks:
-                m["tile"] *= final_dim._max_length
-            final_dim._masks += inner_masks
-            final_dim.axes += dim.axes
-            final_dim.generators += dim.generators
-            final_dim.upper += dim.upper
-            final_dim.lower += dim.lower
-            final_dim._max_length *= dim._max_length
-            final_dim.alternate = final_dim.alternate or dim.alternate
-        return final_dim
+        generators = sum((d.generators for d in dimensions), [])
```

Comment:
Neat! I never knew you could do this... Could do with a comment though.
```suggestion
        # Concatenate generators of all the input dimensions into one list
        generators = sum((d.generators for d in dimensions), [])
```